src/lsm/scan_lookup.zig

```diff
@@ -0,0 +1,228 @@
+const std = @import("std");
+const assert = std.debug.assert;
+
+const Allocator = std.mem.Allocator;
+
+const stdx = @import("../stdx.zig");
+const constants = @import("../constants.zig");
+
+const GridType = @import("../vsr/grid.zig").GridType;
+const ScanType = @import("scan_builder.zig").ScanType;
+
+/// Implements the lookup logic for loading objects from scans.
+pub fn ScanLookupType(comptime Groove: type, comptime Storage: type) type {
+    return struct {
+        const ScanLookup = @This();
+        const Grid = GridType(Storage);
+        const Scan = ScanType(Groove, Storage);
+
+        const Object = Groove.ObjectTree.Table.Value;
+
+        pub const Callback = *const fn (*ScanLookup) void;
+
+        const LookupWorker = struct {
+            scan_lookup: *ScanLookup,
+            lookup_context: Groove.ObjectTree.LookupContext = undefined,
+            index_produced: ?usize = null,
+        };
+
+        groove: *Groove,
+        scan: *Scan,
+        scan_context: Scan.Context = .{ .callback = &scan_read_callback },
+
+        buffer: []Object,
+        buffer_produced_len: usize,
+        state: enum { idle, scan, lookup, finished },
+        callback: ?Callback,
+
+        /// The goal is to fully utilize the disk I/O to ensure the lookup completes as
+        /// quickly as possible, so we run multiple lookups in parallel based on the max
+        /// I/O depth of the Grid.
+        lookup_workers: [Grid.read_iops_max]LookupWorker = undefined,
```

Comment:
Should we apply that to regular prefetch as well? 

https://github.com/tigerbeetle/tigerbeetle/blob/928b36fcf8194acc08a179b7bf5883e28235bfe7/src/lsm/groove.zig#L701-L704

Reply:
Yes, we should -- good catch! And I actually wonder if that will improve the benchmark TPS.

Reply:
It speeds up the scan lookup by a lot!
I found that `512` concurrent scans is the sweet spot for `scan_lookup,` improving the query time by almost 90%:

Before:
```
200 queries in 34.32 s
query latency p00 = 135 ms
query latency p10 = 152 ms
query latency p20 = 156 ms
query latency p30 = 159 ms
query latency p40 = 161 ms
query latency p50 = 163 ms
query latency p60 = 165 ms
query latency p70 = 167 ms
query latency p80 = 170 ms
query latency p90 = 176 ms
query latency p100 = 540 ms
```
After:
```
200 queries in 18.07 s
query latency p00 = 49 ms
query latency p10 = 54 ms
query latency p20 = 61 ms
query latency p30 = 67 ms
query latency p40 = 72 ms
query latency p50 = 82 ms
query latency p60 = 97 ms
query latency p70 = 107 ms
query latency p80 = 113 ms
query latency p90 = 125 ms
query latency p100 = 352 ms
```

For the prefetch lookup, it didn't help that much. In fact, what marginally improved performance was the simplification of the worker logic, breaking the `for` loop if we finish synchronously.

So, I kept the prefetch lookup using `Grid.read_iops_max` and derived a new constant for `ScanLookup.lookups_workers_max`, which makes sense since they have very different workloads.